sdks/python/apache_beam/runners/portability/fn_api_runner.py

```diff
@@ -1142,6 +1202,38 @@ def __init__(self, state, provision_info):
     self.data_server.start()
     self.control_server.start()
 
+  def close(self):
+    self.control_handler.done()
+    to_wait = [
+        self.control_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS),
+        self.data_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS),
+        self.state_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS),
+        self.logging_server.stop(self._DEFAULT_SHUTDOWN_TIMEOUT_SECS)
+    ]
+    for w in to_wait:
+      w.wait()
+
+
+class GrpcWorkerHandler(WorkerHandler):
+  """An grpc based worker_handler for fn API control, state and data planes."""
+
+  def __init__(self, state, provision_info, grpc_server):
+    super(GrpcWorkerHandler, self).__init__(
+        grpc_server.control_handler, grpc_server.data_plane_handler,
+        state, provision_info)
+    self.state = state
+
+    self.control_address = grpc_server.control_address
+    self.control_conn = grpc_server.control_handler \
+      .connections_by_worker_id[self.worker_id]
+
+    self.data_port = grpc_server.data_port
+    self.data_conn = grpc_server.data_plane_handler \
+      .connections_by_worker_id[self.worker_id]
+
+    self.state_port = grpc_server.state_port
+    self.logging_port = grpc_server.logging_port
+
   def data_api_service_descriptor(self):
     return endpoints_pb2.ApiServiceDescriptor(
         url='localhost:%s' % self.data_port)
```

Comment:
Maybe write this as `url='localhost:%s' % self.grpc_server.data_port` rather than storing `data_port` on `self`. Similarly for the others.